sdk/core/azure-core/azure/core/pipeline/policies/_universal.py

```diff
@@ -140,10 +143,16 @@ def on_request(self, request):
         :param request: The PipelineRequest object
         :type request: ~azure.core.pipeline.PipelineRequest
         """
-        request_id = request.context.options.pop('request_id', self._request_id)
-        if request_id:
-            header = {"x-ms-client-request-id":request_id}
-            request.http_request.headers.update(header)
+        if self._auto_request_id:
+            request_id = str(uuid.uuid1())
+        else:
+            request_id = None
+        if self._request_id:
+            request_id = self._request_id
+        if 'request_id' in request.context.options:
+            request_id = request.context.options.pop('request_id')
```

Comment:
If the order of checks is reversed, then these cases can be made exclusive which is a little simpler to reason about (i.e. no "tentatively set but maybe overwrite later"):
```
if 'request_id' in request.context.options:
    request_id = request.context.options.pop('request_id')
elif self._request_id:
    request_id = self._request_id
elif self._auto_request_id:
    request_id = str(uuid.uuid1()) 
else:
    request_id = None
```